### ISSUE TITLE
Get and set the reading reading position of the editor

### ISSUE THREAD

**Issue by FelixMittermeier** [singerdmx/flutter-quill#62]
What is the best way to get (and then later set again) the current position the user has scrolled to? I have very long texts in my app and when the user starts reading, then closes the app and reopens it later again, the app should have "remembered" the reading progress and automatically scroll to the last position.
So in other words I need something like an index of the current visible part of the text.
Such a thing as the scroll position in pixels is not sufficient, because this would only be correct again on the same device, but e.g. wrong on the other devices of the user (the reading progress will be saved in the cloud in order to allow seamless reading across multiple devices).
An answer would be highly appreciated.

**Comment by singerdmx** [comment]
This is done by scrollcontroller which is externally retrievable

**Comment by singerdmx** [comment]
Check apis provided by scrollcontroller
I think it can give what u want

**Comment by ArjanAswal** [comment]
> This is done by scrollcontroller which is externally retrievable 
> 

But it won't work on other devices.
The metadata from scrollcontroller won't give us the pointer to the text line to which the user has read so far.

**Comment by singerdmx** [comment]
Yes, depends on how accurate you want
You want to go back to roughly where it is without focus or you need the focus as well?

**Comment by FelixMittermeier** [comment]
Yeah it has to be quite accurate because the scroll position is completely different if you first open the text on a mobile phone and then later in the browser on a PC with a large monitor or something like that.
Maybe it is possible to count the words which are above the current position? Then it might be possible to scroll to that same position later on other devices too because the words count probably won't change much.

**Comment by singerdmx** [comment]
Internally we track the focus node.
It is possible to leverage that.
All I am saying it is doable.
Just need someone to look into it.

**Comment by ArjanAswal** [comment]
I don't think that there will be a good solution for this problem as platform independence is a requirement you mentioned, and that on itself is a constraint. So the number of pixels you need to scroll over would be different for a phone and a PC and a tablet. 

Also, I am pretty sure that there so way we can extract a pointer to the rendered text that we can store for later use.

However a not-so-neat solution I can come up with would be this :

- Get the total number of characters in the text (n).

- Get the total height of the rendered text (total height you need to scroll to read the entire text) (h).

- Get the height till the text line user has scrolled to while reading (i).

- `(n/h)*i` will roughly give you the _character index_ which the user needs to continue reading from. Store this in cloud.
We will denote it with (c).

Lets say the user opens your app in a PC.

Now, 

- Get the new (h)

- App will retrieve (c) and calculate (i) from `(h*c)/n`

- Use this new height (i) to scroll to.

-  Update (c) on the server

**Comment by ArjanAswal** [comment]
I haven't taken font size and line height into consideration and assumed them to be equal on multiple devices.
Clearly that won't be the usual case so minor tweaks might be needed.

**Comment by FelixMittermeier** [comment]
Hmm this seems way harder than thought.

**Comment by singerdmx** [comment]
You know even QuillJS does not support this

**Comment by singerdmx** [comment]
@FelixMittermeier when you reopen editor are you assuming the same device or not?

**Comment by ArjanAswal** [comment]
> You know even QuillJS does not support this

This is an extremely sophisticated requirement. I can't come up with any word processor which uses this. @FelixMittermeier Do you know of any app which uses this feature?

**Comment by FelixMittermeier** [comment]
@singerdmx  It can be the same but also another device. It depends if the user has multiple devices.
@ArjanAswal For example reading apps like Wattpad save the exact progress. My app is rather similar so this feature is quite critical.

**Comment by ArjanAswal** [comment]
If I am not wrong, Wattpad divides its story into pages and chapters. If you log out, clear the cache and login again, you start from the  beginning of the chapter you were reading.
So simply Wattpad stores the last read chapter and not the text line.

**Comment by singerdmx** [comment]
> It can be the same but also another device. It depends if the user has multiple devices.

meaning a user browses in mobile and reopen it in web, right?
What if user opens the editor in both mobile and web at the same time

**Comment by singerdmx** [comment]
> If I am not wrong, Wattpad divides its story into pages and chapters. If you log out, clear the cache and login again, you start from the beginning of the chapter you were reading.
> So simply Wattpad stores the last read chapter and not the text line.

Yes, it is not arbitrary position. I begin to doubt the feasibility of this feature.

**Comment by FelixMittermeier** [comment]
> > It can be the same but also another device. It depends if the user has multiple devices.
> 
> meaning a user browses in mobile and reopen it in web, right?
> What if user opens the editor in both mobile and web at the same time

This case can be ignored. Then it just uses the later one.

> If I am not wrong, Wattpad divides its story into pages and chapters. If you log out, clear the cache and login again, you start from the beginning of the chapter you were reading.
> Wattpad simply stores the last read chapter and not the text line.

I see, good to know. But they are splitting long texts into very short pages. Might something like that also be possible with quill? I found https://pub.flutter-io.cn/packages/page_turn which looks quite good but I am not sure if it can be used in combination with the editor?

**Comment by ArjanAswal** [comment]
> Yes, it is not arbitrary position. I begin to doubt the feasibility of this feature.

This feature looks like an overkill to me.
@FelixMittermeier I think you should just save the current chapter or page number till which the user has read in the cloud.
HOWEVER persist the exact reading location in the local device, because that is where it matters.

**Comment by singerdmx** [comment]
I think we can hook up something in dispose to remember last read position

**Comment by FelixMittermeier** [comment]
> > Yes, it is not arbitrary position. I begin to doubt the feasibility of this feature.
> 
> This feature looks like an overkill to me.
> @FelixMittermeier I think you should just save the current chapter or page number till which the user has read in the cloud.
> HOWEVER persist the exact reading location in the local device, because that is where it matters.

Okay but what is the best way to use multiple pages with e.g. the page library linked above? Is it still performant if one uses a separate instance of the editor for each page?

**Comment by ArjanAswal** [comment]
> I see, good to know. But they are splitting long texts into very short pages. Might something like that also be possible with quill? I found https://pub.flutter-io.cn/packages/page_turn which looks quite good but I am not sure if it can be used in combination with the editor?

If you want to display text like this which is perfectly possible, you will need to divide your story schema into chunks. Is your app deployed to the app store? Because then it would be impossible as you can't change the back-end database structure while people are using it.

**Comment by ArjanAswal** [comment]
> I think we can hook up something in dispose to remember last read position
> 

That would be very cool. But how would that happen?

**Comment by FelixMittermeier** [comment]
Not yet so I can still change things. Will have look at the page_turn package later then if that should work fine.

**Comment by ArjanAswal** [comment]
> Okay but what is the best way to use multiple pages with e.g. the page library linked above?

Is your app deployed?
If that is the case then all you can change is the front end.

> Is it still performant if one uses a separate instance of the editor for each page?

You don't need to use seperate instance as the user can only read one chapter at a time.

**Comment by FelixMittermeier** [comment]
No it's not yet deployed so that is no problem. Ok just thought multiple instances were needed to render the page turn animation.

**Comment by ArjanAswal** [comment]
> No it's not yet deployed so that is no problem. Ok just thought multiple instances were needed to render the page turn animation.

Oh yes you are right. At most you will be needing 2 instances. Shouldn't be much of an issue performance wise.

**Comment by florianh01** [comment]
I'm unsure why this approach shouldn't work:

The height of the Quill Editor can be calculated [in this way](https://stackoverflow.com/questions/49307677/how-to-get-height-of-a-widget). You can also listen to the [current position](https://api.flutter.dev/flutter/widgets/ScrollController-class.html) of the scroll controller. 

Then, onWillPop divide the two values: {ScrollControllerPixelHeight} / {QuillHeight} = Percentage Value (e.g. 64% read) and safe the percentage to cloud.

On new device: initState of ScrollControllerPosition = 64% * {QuillHeightOnNewDevice}

Did I miss a point here?

**Comment by FelixMittermeier** [comment]
> I'm unsure why this approach shouldn't work:
> 
> The height of the Quill Editor can be calculated [in this way](https://stackoverflow.com/questions/49307677/how-to-get-height-of-a-widget). You can also listen to the [current position](https://api.flutter.dev/flutter/widgets/ScrollController-class.html) of the scroll controller.
> 
> Then, onWillPop divide the two values: {ScrollControllerPixelHeight} / {QuillHeight} = Percentage Value (e.g. 64% read) and safe the percentage to cloud.
> 
> On new device: initState of ScrollControllerPosition = 64% * {QuillHeightOnNewDevice}
> 
> Did I miss a point here?

This approach stops working if new content is added at the end of the chapter. But otherwise it seems actually like a better way to solve it.

**Comment by ArjanAswal** [comment]
> Did I miss a point here?

You are assuming that the entire text fits inside the editor.

**Comment by FelixMittermeier** [comment]
> This is done by scrollcontroller which is externally retrievable

I actually tried this out right now but this does not seem to work either. I created a ScrollController and passed it to the quill editor but whenever I try to get the scroll position with `scrollController.position.pixels`
it just returns 0.0. Did you manage to get a valid position with the scroll controller?

**Comment by singerdmx** [comment]
I guess scrollcontroller is triggered only when you type in the bottom or in the top so that editor needs to scroll one more line to show it

**Comment by singerdmx** [comment]
I have given some thought on this issue and I feel it would be too much to take if you want multiple platform support without tokenizing the content. That is why other editor does not support it.

**Comment by FelixMittermeier** [comment]
Yeah I understand this of course.
The minimum required functionality would just be too save the current scroll position in pixels, save it locally on the device and then reuse this value later (all on the same device). But as mentioned above the scrollController always returns zero and this value would be needed for the minimum functionality.

**Comment by singerdmx** [comment]
I am not sure that can be detected

**Comment by florianh01** [comment]
I have a feeling that the scroll controller is not working properly. I mentioned a similar problem in the Slack group. In my code, the cursor moves under the keyboard with every new line and the scroll controller does nothing. Unfortunately, I haven't had time to look at the code for this yet.

**Comment by FelixMittermeier** [comment]
I ended up wrapping the editor in a listview widget which I passed another (self-managed) scrollController and this works surprisingly well. I can now receive the scrollController.position.pixels value from the listview, save it and reuse it later.
Nevertheless this is more like a workaround - as @florianh01 mentioned too there is something wrong with the internal usage of the scroll controller in the RawEditor. I just do not know what.
But for now I can live with the workaround and will maybe come back again later to improve this.
